### PR TITLE
(obsolete - do not merge) [Inference API] Add Azure AI Studio Embeddings and Chat Completion Support

### PR DESCRIPTION
**Pull request by markjhoy**
This PR adds support for Azure AI Studio integration into the Inference API. Currently this supports `text_embedding` and `completion` task types.

### Prerequisites to Model Creation
* You must have an Azure subscription with [Azure AI Studio](https://ai.azure.com/) access
* You must have a [deployed model](https://ai.azure.com/explore/models) either Chat Completion or Embeddings

image

### Model Creation:
```
PUT _inference/{tasktype}/{model_id}
{
  "service": "azureaistudio",
  "service_settings": {
    "api_key": "{api_key}",
    "target": “{deployment_target}”,
    “provider”: “(model provider}”,
    “endpoint_type”: “(endpoint type)”
  }
}
```
* Valid `{tasktype}` types are: [`text_embedding`, `completion`]

### Required Service Settings:
* `api_key`: The API key can be found on your Azure AI Studio deployment's overview page
* `target`: The target URL can be found on your Azure AI Studio deployment's overview page
* `provider`: Valid provider types are (case insensitive): 
  * `openai` - available for embeddings and completion
  * `mistral` - available for completion only
  * `meta` - available for completion only
  * `microsoft_phi` - available for completion only
  * `cohere` - available for embeddings and completion
  * `snowflake` - available for completion only
  * `databricks` - available for completion only
* `endpoint_type`: Valid endpoint types are:
  * `token` - a "pay as you go" endpoint (charged by token)
    * Available for OpenAI, Meta and Cohere
  * `realtime` - a realtime endpoint VM deployment (charged by the hour)
    * Available for Mistral, Meta, Microsoft Phi, Snowflake and Databricks

### Embeddings Service Settings
* `dimensions`: (optional) the number of dimensions the resulting output embeddings should have. 

### Embeddings Task Settings
(this is also overridable in the inference request)
* `user`: (optional) a string that is a unique identifier representing your end-user. This helps Azure AI Studio in the case of abuse or issues for debugging.

### Completion Service Settings
(no additional service settings)

### Completion Task Settings
(these are all optional and can be overridden in the inference request)
* `temperature`: What sampling temperature to use, between 0 and 2. Higher values mean the model takes more risks. Try 0.9 for more creative applications, and 0 (argmax sampling) for ones with a well-defined answer. Microsoft recommends altering this or top_p but not both.
* `top_p`: An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered. Microsoft recommends altering this or temperature but not both.
* `do_sample`: request to perform the sampling or not
* `max_new_tokens`: the maximum number of new tokens the chat completion inference should produce in the output

### Text Embedding Inference
```
POST _inference/text_embedding/{model_id}
{
    "input": "The answer to the universe is"
}
```

### Chat Completion Inference
```
POST _inference/completion/{model_id}
{
    "input": "The answer to the universe is"
}
```

### CODE REVIEW

**Review by davidkyle — COMMENTED** [review]
Looks great

It's a very large PR I'd like to do some testing then take another look

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/AzureAiStudioEmbeddingsRequestManager.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.elasticsearch.xpack.inference.external.response.azureaistudio.AzureAiStudioEmbeddingsResponseEntity;
+import org.elasticsearch.xpack.inference.services.azureaistudio.embeddings.AzureAiStudioEmbeddingsModel;
+
+import java.util.List;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.xpack.inference.common.Truncator.truncate;
+
+public class AzureAiStudioEmbeddingsRequestManager extends AzureAiStudioRequestManager {
+    private static final Logger logger = LogManager.getLogger(AzureAiStudioChatCompletionRequestManager.class);
```
[the remark this answers is not in the document]
Comment (markjhoy) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/AzureAiStudioEmbeddingsRequestManager.java:31]: Ah - nice catch. The joys of similar base code that's copied and pasted :)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/ChatCompletionResponseEntity.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import org.elasticsearch.inference.InferenceServiceResults;
+import org.elasticsearch.xpack.core.inference.results.ChatCompletionResults;
+import org.elasticsearch.xpack.inference.external.http.HttpResult;
+import org.elasticsearch.xpack.inference.external.http.retry.ResponseParser;
+import org.elasticsearch.xpack.inference.external.request.Request;
+
+import java.io.IOException;
+
+/**
+ * TODO - fill in comment
```
[the remark this answers is not in the document]
Comment (markjhoy) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/ChatCompletionResponseEntity.java:19]: Ooof... I did a "Find All" to look for any `TODO` markers I had left... looks like I missed this one ;)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureaistudio/AzureAiStudioEndpointType.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public String toString() {
+        return name().toLowerCase(Locale.ROOT);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeEnum(this);
```
[the remark this answers is not in the document]
Comment (markjhoy) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureaistudio/AzureAiStudioEndpointType.java:49]: Ah - that's a good idea... I'll take a look

[the remark this answers is not in the document]
Comment (markjhoy) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureaistudio/AzureAiStudioEndpointType.java:49]: > and use the read/writeEnum methods.
🤦  - I realized that in the ServiceSettings superclass, I am actually using these methods, so the `Writable` implementation is not needed on either this or the `Provider` enums...

**Comment by markjhoy** [comment]
buildkite test this

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureaistudio/AzureAiStudioService.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        Model model,
+        List<String> input,
+        Map<String, Object> taskSettings,
+        InputType inputType,
+        TimeValue timeout,
+        ActionListener<InferenceServiceResults> listener
+    ) {
+        var actionCreator = new AzureAiStudioActionCreator(getSender(), getServiceComponents());
+
+        if (model instanceof AzureAiStudioEmbeddingsModel embeddingsModel) {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureaistudio/AzureAiStudioService.java:74]: I mentioned this above, but the reason I chose the visitor pattern for the other services was to let the model dictate the action that is used. That way we don't need the `instanceof` checks here, except one to ensure that it's a `AzureAiStudioModel` model with an `accept()` method.

**Comment by markjhoy** [comment]
Oh - I sometimes hate git and it's ability to not allow a clean merge... going to close this and re-open a new one that's clean.

**Review by jonathan-buttner — COMMENTED** [review]
I'm going to add these here so I don't lose them haha, I'll copy them over to the new PR when it's up 👍

**Comment by markjhoy** [comment]
Sorry @jonathan-buttner ! :) -- new PR is up here: https://github.com/elastic/elasticsearch/pull/108472

**Pull request closed without merging**